### PR TITLE
Network Overview

### LINKED ISSUE #39: NetworkMap
Network map is a summary of access points and connected devices within the network.

Overview
image
https://share.goabstract.com/30871437-23aa-453d-bb0c-96b8a1be9a4d?mode=design&sha=c609991f3db28c212d5b7a820c3a81423b3681a6
Overview is a summary of APs and connected devices within the network.
It shows how many APs and devices tere are in total as well as a list of access points which when are clicked they redirect to their 'detail' route.
Each of the devices have a status indicator saying whether a device is offline or online.

Detail
image
https://share.goabstract.com/546fb614-65a0-410c-8490-cc3f3d20d608?mode=design&sha=c609991f3db28c212d5b7a820c3a81423b3681a6

### Questions
- [ ] What exactly the device colour status mean?
  - [ ] What green stands for?
  - [ ] What orange stands for?
  - [ ] What red stands for?
  Does this relate to they 'health' in percents?
  Would offline or 0 - 10, 10 - 70, 80 - 100 ranges for red, orange, green respectively make sense?
  https://github.com/Telecominfraproject/wlan-cloud-ucentralgw/blob/master/openapi/ucentral/ucentral.yaml#L1727
- [ ] Where do we get the data from?
  - [ ] Devices list https://github.com/Telecominfraproject/wlan-cloud-ucentralgw/blob/master/openapi/ucentral/ucentral.yaml#L1727
  https://github.com/Telecominfraproject/wlan-cloud-ucentralgw/blob/master/openapi/ucentral/ucentral.yaml#L76
  - [ ] Access points or other devices making the mesh seem to be easily identifiable by their `deviceType=AP | SWITCH | MESH | IOT`, How to identify devices connected to a particular access point, switch...?
- [ ] Given there are multiple device types, are we going to show other devices than switch? If so, should we have different icons for them?

### Resources
- CLI, can also be used to inspect requests
There's package with sample requests and cli commands available to download from slack
https://ucentral-ui.slack.com/archives/C020QQ6KQ06/p1622010151002800
- OpenAPI
https://github.com/Telecominfraproject/wlan-cloud-ucentralgw/blob/master/openapi/ucentral/ucentral.yaml#L1777

### TODOS
- [x] Create network_map and network_map/:id routes
- [x] Allow for redirecting from 'index' to network_map.
- [x] Mock 'list devices' request with multiple 'deviceTypes'
- [x] Show devices count with `deviceType=AP`
- [x] Show all devices count
- [x] On AP list show their SSIDs, health status and amount of connected devices.
- [x] Create device tiles using Uc::LinkTile which once are clicked they redirect to `network_map/:deviceId` route.
- [x] Allow for Uc::LinkTile to show device health status.
Perhaps creating another component that's similar to Uc::LinkTile might be a better solution so not to create a too big component
- [x] On `network_map/:deviceId` route - Access point details route, show a list of devices that are connected to it.
  - remaining question is, how do we know that a device is connected to this particular access point?
- [x] On the devices list show their names, mac addresses and Ip addresses.
- [x] Since for this task we're data heavy, it'd be best to define ember-cli-mirage factories for testing and development.

### PR DESCRIPTION
**Pull request by BobrImperator**
closes #39 - needs Back Navigation from #36 which I will split in a moment

- Creates an http service to make requesting resources a bit easier
for now it only implements GET with support for query params and headers
- Creates network overview with summary of devices and current APs
- Creates NetworkOverview/detail with details for a specifc AP picked from a list on Network Overview
- Creates a component to show different device icons and their health statuses
- passes `...attributes` to `Uc::Layout` to allow for module specific styles
- Adds handling of `@model` argument for `LinkTitle`
- Adds `:icon` named block check in `LinkTitle` so `LinkTitle` allows for passing custom `icon` component i.e `Uc::DeviceWithStatus`.

### CODE REVIEW

File: app/components/network-map.hbs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  <:title>
+    {{t "network_map.title"}}
+  </:title>
+
+  <:content>
+    <div
+      local-class="summary"
+      {{did-insert (perform this.loadAccessPointsTask)}}
+      {{did-insert (perform this.loadDevicesCountTask)}}
```
Comment (marcoow) [inline app/components/network-map.hbs:14]: Why aren't we doing these operations in the route? That should make things simpler and improve UX since we wouldn't first show 0 devices only to update that to the real number a moment later?

File: app/components/network-map/detail.css [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  @apply inline-block align-middle;
+}
+
+.layout {
+  padding: 0 !important;
+}
+
+.layout-navigation {
+  @apply p-4 pb-0 inline-block self-start;
+}
```
Comment (marcoow) [inline app/components/network-map/detail.css:54]: Why is this style necessary? In the designs the back arrow looks the same everywhere it appears?

File: app/components/network-map/detail.hbs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    {{t "network_map.detail.title"}}
+    <p local-class="location-title">
+     {{this.loadDataTask.lastSuccessful.value.accessPoint.location}}
+    </p>
+  </:title>
+
+  <:content>
+    <div
+      local-class="summary"
+      {{did-insert (perform this.loadDataTask)}}
```
Comment (marcoow) [inline app/components/network-map/detail.hbs:18]: See above – I'm not sure why we do this in the component vs. in the route?

File: app/components/network-map/detail.css [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+.device-list__item-icon {
+  @apply inline-block ml-5 mr-4 align-middle;
+}
+
+.device-list__item-name {
+  @apply inline-block align-middle;
+}
+
+.layout {
+  padding: 0 !important;
```
Comment (marcoow) [inline app/components/network-map/detail.css:49]: Why do we need this? Aren't all pages supposed to use the same layout?

File: app/routes/network-map/show.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import Route from "@ember/routing/route";
+import { inject as service } from "@ember/service";
+
+export default class NetworkMapShowRoute extends Route {
+  @service session;
+
+  beforeModel(transition) {
+    this.session.requireAuthentication(transition, "auth");
+  }
```
Comment (marcoow) [inline app/routes/network-map/show.js:9]: This and the same code in `app/routes/network-map/index.js` could also be moved into `app/routes/network-map.js` which would allow deleting these extra files.

File: app/services/http.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    });
+  }
+
+  createUrl(urlString, queryParams = {}) {
+    const url = new URL(urlString, ENV.APP.BASE_API_URL);
+
+    for (const [key, value] of Object.entries(queryParams)) {
+      if (value === null || value === undefined) {
+        // noop
+      } else {
```
Comment (marcoow) [inline app/services/http.js:22]:
```suggestion
      if (value !== null && value !== undefined) {
```

;)

**Review by marcoow — COMMENTED** [review]
I'm not quite sure why we're loading the data in the component vs. in the routes? There doesn't seem to be an obvious benefit in this case?

Also I was wondering: do the mocked responses to the requests loading the network info resemble the actual responses we'll be getting from the API?

**Comment by BobrImperator** [comment]
@marcoow routes are blocking unless a POJO is returned which in turn is confusing. It should also be easier for others new to ember to manage data an reloading if needed without this weird routes behavior.

File: app/components/network-map/detail.css (hunk 2 above)
Comment (BobrImperator) [reply app/components/network-map/detail.css:54]: For some reason its appearing in the middle for that view, and also it needs to have some padding that had to be reset because detail view is a bit different than others.

Comment (marcoow) [reply app/components/network-map/detail.css:54]: I'm not following – all designs in Abstract show it at the top left? https://share.goabstract.com/e224b8d6-8117-4698-b54c-1699bad6526b?sectionId=acd1bc40-e1c1-4c88-ae70-194b2907ebef

Comment (BobrImperator) [reply app/components/network-map/detail.css:54]: I meant to say that it was appearing in the middle of the screen when wrapped in a div, which I have to add due to the detail view having an items list that has to cover the whole width of a device while the base layout adds padding.

This is why in some other line there's resetted padding, and here the padding only for the pagination is added again.

Comment (marcoow) [reply app/components/network-map/detail.css:54]: Can we not change the navigation component so that it works for all screens without additional styles for individual screens?

Comment (BobrImperator) [reply app/components/network-map/detail.css:54]: I don't think so, because navigation component isn't the only problem here, but the whole layout.
It also doesn't make sense to change the basic components just because 1 screen in the whole app is a bit different.

File: app/components/network-map/detail.css (hunk 4 above)
Comment (BobrImperator) [reply app/components/network-map/detail.css:49]: Only details is a bit different because of its device list spanning through the whole width of a device.

File: app/components/network-map/detail.css (hunk 2 above)
Comment (marcoow) [reply app/components/network-map/detail.css:54]: I think I still don't understand what's different in terms of layout here. Can we pair on it real quick some time today?

File: app/components/network-map/detail.css (hunk 4 above)
Comment (marcoow) [reply app/components/network-map/detail.css:49]: Ah I see – that's this right? https://share.goabstract.com/e224b8d6-8117-4698-b54c-1699bad6526b?collectionLayerId=b08951e1-921b-4151-9cfe-220a605a1e4a&mode=design

I wonder whether we could have a better mechanism to handle this – maybe some kind of `<Uc::Layout::ContentWrapper>` component (similar to e.g. `<Uc::Layout::Logo>`) that adds the padding and that we'd simply **not** use in this case?

File: app/components/network-map/detail.css (hunk 2 above)
Comment (marcoow) [reply app/components/network-map/detail.css:54]: Ok, think I understand now – this is necessary because of https://github.com/blogic/ucentral-pwa/pull/50#discussion_r650933275, right? I think there should be a simpler way to address this – happy to help exploring a way!

File: app/components/network-map.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      const response = yield this.http.get("/api/v1/devices", {
+        countOnly: true,
+      });
+
+      return yield response.json();
+    } catch (error) {
+      console.error(error);
+    }
+  }
+}
```
Comment (marcoow) [inline app/components/network-map.js:33]: This file doesn't seem to be needed at all?

File: app/components/uc/link-tile.hbs [hunk 8]
```diff
@@ -1,8 +1,27 @@
...
+    {{else}}
+      <div local-class="icon">
+        {{inline-svg @icon}}
+      </div>
+    {{/if}}
+    <p local-class="text">
+      {{yield}}
+    </p>
+  </LinkTo>
+{{/if}}
```
Comment (marcoow) [inline app/components/uc/link-tile.hbs:27]: Is duplicating all of the component necessary to distinguish between `@model` being passed and it not being passed?

File: app/templates/index.hbs [hunk 9]
```diff
@@ -1 +1 @@
-<Dashboard />
\ No newline at end of file
+<Dashboard />
```
Comment (marcoow) [inline app/templates/index.hbs:1]: seems unrelated?

**Review by marcoow — COMMENTED** [review]
Can you move the `content.Wrapper` change into a separate PR that we merge before this?

File: app/components/uc/link-tile.hbs (hunk 8 above)
Comment (BobrImperator) [reply app/components/uc/link-tile.hbs:27]: Yes, unfortunately since the LinkTo is a wrapper here, we end up copying all other stuff too.

We've discussed it recently on slack, please take a look https://simplabs.slack.com/archives/C55DR890R/p1623333838003600

Comment (marcoow) [reply app/components/uc/link-tile.hbs:27]: I see 😞

File: config/environment.js [hunk 10]
```diff
@@ -23,6 +23,7 @@ module.exports = function (environment) {
       AUTHENTICATION_URL:
         "https://ucentral.dpaas.arilia.com:16001/api/v1/oauth2",
       DEVICE_URL: "https://ucentral.dpaas.arilia.com:16001/api/v1/device",
+      BASE_API_URL: "https://ucentral.dpaas.arilia.com:16001",
```
Comment (marcoow) [inline config/environment.js:26]: This seems confusing – should we not change `AUTHENTICATION_URL` and `DEVICE_URL` to no longer include `BASE_API_URL` then? This also seems like a perfect thing to move out into a separate PR that goes before this.

**Review by marcoow — COMMENTED** [review]
Sorry, saw one more thing I think we should address. Let's extract that and address the other open conversation, rebase this and get it merged 👍

**Review by marcoow — APPROVED** [review]
